thrust/thrust/tuple.h

```diff
@@ -125,3 +147,34 @@ using ::cuda::std::tie;
  */
 
 THRUST_NAMESPACE_END
+
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+template<>
+struct tuple_size<tuple<THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type>> : tuple_size<tuple<>> {};
+
+template<class T0>
+struct tuple_size<tuple<T0, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type>> : tuple_size<tuple<T0>> {};
+
+template<class T0, class T1>
+struct tuple_size<tuple<T0, T1, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type>> : tuple_size<tuple<T0, T1>> {};
+
+template<class T0, class T1, class T2>
+struct tuple_size<tuple<T0, T1, T2, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type>> : tuple_size<tuple<T0, T1, T2>> {};
+
+template<class T0, class T1, class T2, class T3>
+struct tuple_size<tuple<T0, T1, T2, T3, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type>> : tuple_size<tuple<T0, T1, T2, T3>> {};
+
+template<class T0, class T1, class T2, class T3, class T4>
+struct tuple_size<tuple<T0, T1, T2, T3, T4, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type>> : tuple_size<tuple<T0, T1, T2, T3, T4>> {};
+
+template<class T0, class T1, class T2, class T3, class T4, class T5>
+struct tuple_size<tuple<T0, T1, T2, T3, T4, T5, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type>> : tuple_size<tuple<T0, T1, T2, T3, T4, T5>> {};
+
+template<class T0, class T1, class T2, class T3, class T4, class T5, class T6>
+struct tuple_size<tuple<T0, T1, T2, T3, T4, T5, T6, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type>> : tuple_size<tuple<T0, T1, T2, T3, T4, T5, T6>> {};
+
+template<class T0, class T1, class T2, class T3, class T4, class T5, class T6, class T7>
+struct tuple_size<tuple<T0, T1, T2, T3, T4, T5, T6, T7, THRUST_NS_QUALIFIER::null_type>> : tuple_size<tuple<T0, T1, T2, T3, T4, T5, T6, T7>> {};
```

Comment:
A bit late to the party, but: why does this only go up to 8 non-null elements (0-7)? Shouldn't this go up to 9 (0-8)? AFAICT the old tuple implementation supported up to 10 non-null types, so 9 non-null types plus 1 null type should be handled here?

If true, I think this was unfortunately hidden by the testing added in 471b220d5b46bd4a3163659f532fdeecb6a71d6d falling one element short, as well. There's no test for the case of 9 non-null types followed by `thrust::null_type` having a tuple size of 9. I believe this would currently (incorrectly) produce the standard tuple size of 10.

Reply:
Thinking further, I'm guessing that these specializations were added for compatibility with existing code that fully parameterizes `tuple`? In that case, since they would be specifiying all 10 parameters and all of these specializations only specify 9, none these specializations would match.

When I made my initial comment, I thought that, while only specifying 9 parameters for all of these specializations seemed odd, it was ok because the template parameter defaults specified for `tuple` would provide `thrust::null_type` for the 10th. But that was only true of the old tuple definition, in which case this compatibility layer wouldn't be necessary anyway.

tl;dr: In addition to adding the 9 + 1 specialization, all existing specializations should be extended from 9 to 10 template parameters.

Reply:
😱 Seems like some of them did not make it